File: data-structure/map/map.js
class Map {
    constructor() {
        this.entries = [];
        this.size = 0;
        this.capacity = 10;
    }

    // 重新分配容量
    resizeMap(newCapacity) {
        this.capacity = newCapacity;
    }

    // 插入键值对（如果存在则更新）
    put(key, value) {
        for (let i = 0; i < this.size; i++) {
            if (this.entries[i].key === key) {
                this.entries[i].value = value; // 更新值
                return;
            }
        }
        if (this.size >= this.capacity) {
            this.resizeMap(this.capacity * 2);
        }
        this.entries[this.size] = {
            key,
            value
        };
        this.size++;
    }

    // 查找键
    get(key) {
        for (let i = 0; i < this.size; i++) {
            if (this.entries[i].key === key) {
                return this.entries[i].value;
            }
        }
        return -1; // 未找到
    }

    // 删除键
    delete(key) {
        for (let i = 0; i < this.size; i++) {
            if (this.entries[i].key === key) {
                this.entries.splice(i, 1);
                this.size--;
                return;
            }
        }
    }
}

// 测试
const map = new Map();

map.put("apple", 10);
map.put("banana", 20);
map.put("orange", 30);

console.log("apple:", map.get("apple"));
console.log("banana:", map.get("banana"));
console.log("grape:", map.get("grape"));

map.delete("banana");
console.log("banana after delete:", map.get("banana"));

/*
jarry@MacBook-Pro map %  node map.js 
apple: 10
banana: 20
grape: -1
banana after delete: -1
*/

File: data-structure/map/map.py
class Map:
    def __init__(self):
        self.entries = []
        self.size = 0
        self.capacity = 10

    # 重新分配容量
    def resize_map(self, new_capacity):
        self.capacity = new_capacity

    # 插入键值对（如果存在则更新）
    def put(self, key, value):
        for entry in self.entries:
            if entry['key'] == key:
                entry['value'] = value  # 更新值
                return
        if self.size >= self.capacity:
            self.resize_map(self.capacity * 2)
        self.entries.append({'key': key, 'value': value})
        self.size += 1

    # 查找键
    def get(self, key):
        for entry in self.entries:
            if entry['key'] == key:
                return entry['value']
        return -1  # 未找到

    # 删除键
    def delete(self, key):
        for i in range(self.size):
            if self.entries[i]['key'] == key:
                self.entries.pop(i)
                self.size -= 1
                return


# 测试
map = Map()

map.put("apple", 10)
map.put("banana", 20)
map.put("orange", 30)

print("apple:", map.get("apple"))
print("banana:", map.get("banana"))
print("grape:", map.get("grape"))

map.delete("banana")
print("banana after delete:", map.get("banana"))

"""
jarry@MacBook-Pro map % python map.py 
apple: 10
banana: 20
grape: -1
banana after delete: -1
"""
